docs/design-docs/01-oci-support.md

```diff
@@ -0,0 +1,338 @@
+
+### `kpt pkg get` sub-packages
+
+It is possible to use `kpt pkg get` to add sub-packages to a target location.
+
+Syntax for a sub-package target location is unchanged, it's a normal filesystem path.
+
+Syntax for an OCI sub-package source location requires the ability to tell when an image name ends and a sub-package path inside that image begins. In `git` this requires an explicit `.git` extension at the transition, and in `.oci` this requires double slash.
+
+```shell
+# clone sub-package as new sub-folder
+kpt pkg get oci://us-docker.pkg.dev/the-project-id/the-repo-name/the-package//simple/example:v3 my-package/simple/my-example
+```
+
+### `kpt pkg update`
+
+The command for update is not changed, but when the `upstream` is `oci` then the `@VERSION` is used to change the `upstream` image's `tag` or `digest` value.
+
+`kpt pkg update @v14` and `kpt pkg update DIR@v14` will assign `:v14` as the new upstream image tag.
+
+`kpt pkg update @sha256:xxxxx` and `kpt pkg update DIR@sha256:xxxxx` will assign `@sha256:xxxxx` as the new upstream image digest.
+
+`kpt pkg update` and `kpt pkg update DIR` use the currently defined `upstream` image tag or digest unchanged.
+
+At that point, if the `upstream` is an `image:tag` that is to discover the current `image:digest` for tag, otherwise the `upstream` value for `image:digest` is used. In either case, the `upstreamLock` is changed to point at that new `image:digest`. 
+
+The package contents of the old and new `upstreamLock` image digest are fetched to temp folders, and are the basis of the 3-way merge to update the target package.
+
+### `kpt pkg diff`
+
+The `kpt pkg diff` command is identical to `kpt pkg update` in the way that `[PKG_PATH@VERSION]` argument is mapped to OCI concepts.
+
+### Command additions
+
+### ` kpt pkg push`
+
+Summary TBD
+
+### ` kpt pkg pull`
+
+Summary TBD
+
+## User Guide
+
+### Creating a package respository
+
+Before kpt packages can be pushed and pulled as OCI images, a suitable repository 
+must be created. Google Artifact Registry and Google Container Registry are both
+excellent choices.
+
+```shell
+# Choose names and locations
+LOCATION="us"
+PROJECT_ID="kpt-demo-73823"
+REPOSITORY_NAME="blueprints"
+
+# Base name for any images in this repository
+REPOSITORY="${LOCATION}-docker.pkg.dev/${PROJECT_ID}/${REPOSITORY_NAME}"
+
+# Create the repository
+gcloud artifacts repositories create --location="${LOCATION}" --repository-format=docker --project="${PROJECT_ID}" "${REPOSITORY_NAME}"
+```
+
+### Creating and pushing a new package
+
+Creating a new package is no different. But when ready to publish, the `kpt pkg push` command 
+is used instead of source control operations.
+
+```shell
+# A package in a new directory
+mkdir hello-world
+kpt pkg init hello-world --description="A simple blueprint"
+
+# Store the contents in the repository, tagged as v1
+kpt pkg push hello-world --image=${REPOSITORY}/hello-world:v1
```

Comment:
Added some docs above. Essentially, using only golang modules and no shell-executes, it would tar up the contents of the source folder, and upload the tar as a single-layer image to the target repository.

There is also image metadata which can be assigned at the same time. It would be interesting to consider what `Kptfile` properties should be added as image metadata (like title/description)

A bit like `docker push` - except the input is a simple folder so there's no separate `build` step needed.

Reply:
the `kpt pkg push|pull` and possible alternative `kpt pkg copy` are now described